### PR TITLE
Add MovingMNIST

### PR DESCRIPTION
**Pull request by vballoli**
PR in reference to https://github.com/pytorch/vision/issues/2676. Currently in draft since the dataset train-test split needs discussion. Inputs regarding how to better handle the train-test arguments is appreciated! @vfdev-5 @pmeier @fmassa

### CODE REVIEW

File: torchvision/datasets/mnist.py [hunk 1]
```diff
@@ -435,6 +435,74 @@ def get_int(b: bytes) -> int:
     return int(codecs.encode(b, 'hex'), 16)
 
 
+class MovingMNIST(VisionDataset):
+    """MovingMNIST"""
```
Comment (vfdev-5) [inline torchvision/datasets/mnist.py:439]: Could you please a link (http://www.cs.toronto.edu/~nitish/unsupervised_video/) to the dataset like that 
```
`MNIST <http://yann.lecun.com/exdb/mnist/>`_ Dataset.
```
and define docstring `Args` etc as it is done for other datasets.

File: torchvision/datasets/mnist.py [hunk 2]
```diff
@@ -435,6 +435,74 @@ def get_int(b: bytes) -> int:
     return int(codecs.encode(b, 'hex'), 16)
 
 
+class MovingMNIST(VisionDataset):
```
Comment (vfdev-5) [inline torchvision/datasets/mnist.py:438]: Can't we inherit it from `MNIST` ?

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the draft ! I left few comments

File: torchvision/datasets/mnist.py (hunk 2 above)
Comment (pmeier) [reply torchvision/datasets/mnist.py:438]: If we use this as a video dataset, we shouldn't. We probably need to use `MNIST` to generate the training split though.

File: torchvision/datasets/mnist.py [hunk 3]
```diff
@@ -435,6 +435,74 @@ def get_int(b: bytes) -> int:
...
+
+        with open(os.path.join(self.processed_folder, self.file), 'wb') as f:
+            torch.save((data), f)
+
+    def _check_exists(self) -> bool:
+        return (os.path.exists(os.path.join(self.processed_folder,
+                                            self.file)))
+
+
+    def __getitem__(self, index: int) -> Tuple[Any, Any]:
```
Comment (pmeier) [inline torchvision/datasets/mnist.py:480]: Without a length, the dataset is not iterable.

```suggestion
    def __len__(self) -> int:
        return len(self.data)
        
    def __getitem__(self, index: int) -> Tuple[Any, Any]:
```

File: torchvision/datasets/mnist.py [hunk 4]
```diff
@@ -494,3 +562,9 @@ def read_image_file(path: str) -> torch.Tensor:
     assert(x.dtype == torch.uint8)
     assert(x.ndimension() == 3)
     return x
+
+def read_npy_file(path: str) -> torch.Tensor:
+    with open(path, 'rb') as f:
+        x = torch.tensor(np.load(f))
+    assert(x.dtype == torch.uint8)
+    return x
```
Comment (pmeier) [inline torchvision/datasets/mnist.py:570]: Nit

```suggestion
    return x
    
```

**Review by pmeier — CHANGES_REQUESTED** [review]
Hey @vballoli,

thanks a lot for this PR. We discussed this internally and have some comments on this:

1. `MovingMNIST` is a video dataset, i. e. the individual frames have a strong temporal relation. Currently you are treating it as an image dataset. We should return 4D tensor representing a single video (`channels x depth x height x weight`) rather than an image with multiple channels (`channels x height x weight`). This also means that we shouldn't treat the frames individually for the transforms.
2. You only implemented the test split, since this is the only generated data provided by the author. Since the authors also provide the code to generate the sequences (see for example the [implementation by `tensorflow`](https://github.com/tensorflow/datasets/blob/b9deaa07f567ad3211850a1e5c1ae993a1a03409/tensorflow_datasets/video/moving_sequence.py)), it might be beneficial to also provide a train split. If we do this, we need to decide if we want to generate the sequences ahead-of-time or on-the-fly. Depending on how large the training split will be, the former might be to much on a burden on the memory (even the test split is ~1GB and resides completely in memory).
3. Although the authors split the sequences of two with 10 frames each, this should not be required since `MovingMNIST` is marketed as an unsupervised dataset. This is especially important if we generate the trainings data with a deviating number of frames. IMO we should simply return a single video. Maybe we could provide a (static) method that implements this split by the authors.

Let us know what you think and if you are willing to continue working on this.

**Comment by fmassa** [comment]
To complement @pmeier answer, if you could let the dataset return a single tensor of shape `[num_frames, channels, height, width]` it would be great

**Comment by avijit9** [comment]
@pmeier  Somebody working on closing this PR? If no, I can help.

**Comment by pmeier** [comment]
@avijit9 No, as far as I'm aware, no one is currently working on this. Still, lets wait how #3562 plays out. We are currently not sure which datasets we want to add.

**Comment by yassineAlouini** [comment]
This is a new dataset request and will be added to the new dataset API once it is stable enough. In the meantime, we will close the current PR (don't worry, the code will be reused at some point in the new dataset design to integrate it). @pmeier

**Comment by yassineAlouini** [comment]
Thanks @vballoli for this great contribution and sorry for the long time to get a new reply!


Unfortunately, there is a new datasets API in the works [here](https://github.com/pytorch/vision/issues/5336) and new datasets should follow the new API design. 

That being said, rest assured, there are at least two options:

* Given that you already put in the work in this PR,  a new PR adding this to the new API is more than welcome (don't hesitate to ask for help, for example @pmeier is very knowledgable). As a starting point, you can refer to [this guide](https://github.com/pytorch/vision/blob/main/torchvision/prototype/datasets/_builtin/README.md) on how to do that.

* Someone else could take this over and build on what you have done. You will receive proper credit of course. :+1: 

Hope one of the two options works for you @vballoli. :ok_hand:

**Pull request closed without merging**